runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -156,6 +154,172 @@ static TR::Node *lowerCASValues(
    }
 
 
+/**
+ * @brief Add checks to skip (fast-path) acmpHelper call
+ * 
+ * @details
+ * 
+ * This transformation adds checks for the cases where the acmp can be performed
+ * without calling the VM helper. The trasformed Trees represen the following operation:
+ *
+ * 1. If the address of lhs and rhs are the same, produce an eq (true) result
+ *    and skip the call (note the two objects must be the same regardless of
+ *    whether they are value types are reference types)
+ * 2. Otherwise, do VM helper call
+ *
+ * The transformation looks as follows:
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |treetop               |
+ *  |  icall acmpHelper    |
+ *  |    aload lhs         |
+ *  |    aload rhs         |
+ *  |ificmpeq --> ...      |
+ *  |  ==> icall           |
+ *  |  iconst 0            |
+ *  |BBEnd                 |
+ *  +----------------------+
+ *
+ *  ...becomes...
+ *
+ *  +----------------------+
+ *  |ttprev                |
+ *  |iRegStore x           |
```

Comment:
Why does this need to be set to 1 to go to an if test it for equality to 0? Why not just branch directly to the block targetted by the ificmpqeq in the block you are going to...

Reply:
I'm not actually sure that ificmpeq/ificmpne will be there. That's the shape of the trees I'm currently expecting to come out of IlGen. But, I'm not sure I can rely on that shape persisting through to the code generator. I wrote the code so it assumes nothing about how the result of the comparison is used.

If we can guarantee that the comparison will be there, then I agree it would be better to just jump to its target block, as appropriate.

Reply:
We likely need to handle both cases - similar to arraycmp and ifarraycmp

Reply:
What arraycmp and ifarraycmp code are you referring to? I can't seem to find it.

Reply:
The reason you can't find it seems to be it was removed as part of code clean up in OMR. That removal seems to be incorrect so I'm going to create an item to restore it but VMifArrayCmpEvaluator is what it was. The basic point of this was that because arraycmp returns 1/0/2 and the default compare sequences on x86 generate -1/0/1 we could save instructions if we knew we were just going to branch on the result by not generating the correct 1/0/2 but rather just directly testing the CPU flags at the end of the compare sequence.

My point is that generating a register for the result is going to be wasteful if we are just going to test and branch - the sequence should, ideally, either a) result in the flags naturally being set correctly to allow the branch to happen without additional tests or b) we need a pattern match for an if against one of these helpers so that we do set the flags and can jump directly without another test/compare instruction.

If you could create a backlog item for this that would be great. If you won't get to it in the short term you could mark it help wanted to see if anyone else would like to pick it up before you get back to it.

Reply:
I've opened ##9856 to track this work.